### PR TITLE
feat: support Estimator early stopping hooks [DET-3239]

### PR DESCRIPTION
**Pull request by brain-good**
## Test Plan
- [x] tested with a custom hook that calls `run_context.request_stop()` for train spec
- [x] tested with a custom hook that calls `run_context.request_stop()` for val spec

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/estimator/_estimator_trial.py [hunk 1]
```diff
@@ -504,13 +522,20 @@ def _init_model(self) -> None:
         self.train_spec = tf.estimator.TrainSpec(
             input_fn=repeating_train_fn, hooks=self.train_hooks
         )
+        steps = (
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_trial.py:525]: praise: nice job catching & fixing the issue in the case where users do specify `steps` as of their EvalSpec.

File: harness/determined/estimator/_estimator_trial.py [hunk 2]
```diff
@@ -504,13 +522,20 @@ def _init_model(self) -> None:
...
+            self.val_spec.steps // self.context.distributed.get_size()
+            if self.val_spec.steps is not None
+            else None
+        )
         self.eval_spec = tf.estimator.EvalSpec(
-            input_fn=self.val_spec.input_fn, hooks=self.val_spec.hooks, steps=None
+            input_fn=self.val_spec.input_fn, hooks=self.val_hooks, steps=steps
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_trial.py:531]: non-blocking: just to reduce number of class members could we do:

```
self.eval_spec = tf.estimator.EvalSpec(
      input_fn=self.val_spec.input_fn, hooks=self. _init_val_hooks(), steps=steps
```

I would probably do the same for train hooks.

File: harness/determined/estimator/_estimator_trial.py [hunk 3]
```diff
@@ -49,6 +49,23 @@
 VERY_LARGE_NUMBER = 9999999999999999
 
 
+class DeterminedEarlyStoppingHook(tf.compat.v1.train.SessionRunHook):  # type: ignore
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_trial.py:52]: non-blocking: we should add a unit test for this. Something along the lines of having a callback that stops training after and evaluation and then use our checkpointing callback to make sure that everything executes as expected.

File: harness/determined/estimator/_estimator_trial.py (hunk 2 above)
Comment (brain-good) [reply harness/determined/estimator/_estimator_trial.py:531]: I'll do this as a separate PR

**Pull request merged**